sdks/universal-router-sdk/src/swapRouter.ts

```diff
@@ -82,14 +83,30 @@ export abstract class SwapRouter {
     positionManagerOverride?: string
   ): MethodParameters {
     const v4Pool: V4Pool = options.outputPosition.pool
-    const token0 = options.inputPosition.pool.token0
-    const token1 = options.inputPosition.pool.token1
+    let token0: Currency = options.inputPosition.pool.token0
+    let token1: Currency = options.inputPosition.pool.token1
     const v4PositionManagerAddress =
       positionManagerOverride ?? CHAIN_TO_ADDRESSES_MAP[v4Pool.chainId as SupportedChainsType].v4PositionManagerAddress
 
+    const nativeCurrency = Ether.onChain(v4Pool.chainId)
+
+    // if migrating wrapped native -> native
+    if (nativeCurrency.wrapped.equals(token0) || nativeCurrency.wrapped.equals(token1)) {
+      // v4 pool must use native currency and must have useNative flag set
+      invariant(v4Pool.token0.isNative === options.v4AddLiquidityOptions.useNative, 'NATIVE_REQUIRED')
+      // set token0 or token1 to native currency to validate parameters later
+      if (nativeCurrency.wrapped.equals(token0)) {
+        token0 = nativeCurrency
+      } else if (nativeCurrency.wrapped.equals(token1)) {
+        token1 = token0
+        token0 = nativeCurrency
```

Comment:
this seems unnecessary? we already validate that the pool must be a native pool, which means token0 will already be the nativeCurrency

Reply:
oh I see its just to make the invariant pass since its checking the INPUT tokens. we should just write a different invariant. In this case its not always true that input.token0 =/= output.token0 || input.token1 =/output.token1, 

so we should just change the invariant to handle this case!

it should be something like 
wrappedToken = input.token0 == native.wrapped ? input.token0 : input.token1
invariant(wrappedToken == native.wrapped && input.nonWrappedToken == output.token1)

Reply:
isnt token0 always native and never token1? because native==address(0)?

Reply:
but in v3, weth could be token0 or token1. so if weth is token1, then we need to make sure that token0 in v3 = token1 in v4

Reply:
righttt my confusion comes from [this naming](https://github.com/Uniswap/sdks/pull/279#discussion_r1939918631), proof we need to rename i guess 😆